src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -1309,85 +1311,169 @@ public static unsafe int LastIndexOfAny(ref byte searchSpace, byte value0, byte
 
         // Optimized byte-based SequenceEquals. The "length" parameter for this one is declared a nuint rather than int as we also use it for types other than byte
         // where the length can exceed 2Gb once scaled by sizeof(T).
-        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
-        public static unsafe bool SequenceEqual(ref byte first, ref byte second, nuint length)
+        public static bool SequenceEqual(ref byte first, ref byte second, nuint length)
         {
-            IntPtr offset = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
-            IntPtr lengthToExamine = (IntPtr)(void*)length;
-
-            if ((byte*)lengthToExamine >= (byte*)sizeof(UIntPtr))
+            // Use nint for arithmetic to avoid unnecessary 64->32->64 truncations
+            if (length >= sizeof(nuint))
             {
-                // Only check that the ref is the same if buffers are large, and hence
-                // its worth avoiding doing unnecessary comparisons
-                if (Unsafe.AreSame(ref first, ref second))
-                    goto Equal;
+                // Conditional jmp foward to favor shorter lengths. (See comment at "Equal:" label)
+                // The longer lengths can make back the time due to branch misprediction
+                // better than shorter lengths.
+                goto Longer;
+            }
 
-                if (Vector.IsHardwareAccelerated && (byte*)lengthToExamine >= (byte*)Vector<byte>.Count)
+#if TARGET_64BIT
+            // On 32-bit, this will never be true since sizeof(nuint) == 4
+            if (length >= sizeof(uint))
+            {
+                nint offset = 0;
```

Comment:
Just out of curiosity, what's the performance if this block were to be rewritten as such?

```cs
uint differentBits = LoadUInt(ref first) - LoadUInt(ref second);
differentBits |= LoadUInt(ref first, length - 4) - LoadUInt(ref second, length - 4);
return (differentBits == 0);
```

I realize it performs 4 reads instead of 2 in the case where _length = 4_, but maybe the lack of jumps would make up for it? In an ideal world the JIT could optimize it heavily:

```asm
mov tmp, dword ptr [first]
sub tmp, dword ptr [second]
mov tmp2, dword ptr [first + length - 4]
sub tmp2, dword ptr [second + length - 4]
or tmp2, tmp1
setnz al ; reuse flags from 'or' operation above
movzx eax, al
ret
```

Reply:
Might do the `length - 4` to a local first as the JIT doesn't compound it in the address and uses 2 registers
 ```asm
       mov      eax, dword ptr [rcx]
       sub      eax, dword ptr [rdx]
       add      r8, -4
       mov      r9, r8
       mov      ecx, dword ptr [rcx+r9]
       sub      ecx, dword ptr [rdx+r8]
       or       eax, ecx
```

Reply:
```asm
       add      r8, -4
       mov      eax, dword ptr [rcx]
       sub      eax, dword ptr [rdx]
       mov      ecx, dword ptr [rcx+r8]
       sub      ecx, dword ptr [rdx+r8]
       or       eax, ecx
       test     eax, eax
       sete     al
       movzx    rax, al
```